stl/inc/algorithm

```diff
@@ -6977,6 +7125,82 @@ void nth_element(_ExPo&&, _RanIt _First, _RanIt _Nth, _RanIt _Last) noexcept /*
+
+        // clang-format off
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<_It, _Pr, _Pj>
+        constexpr _It operator()(_It _First, _It _Nth, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Nth);
+            _Adl_verify_range(_Nth, _Last);
+            auto _UNth   = _Get_unwrapped(_Nth);
+            auto _UFinal = _Get_final_iterator_unwrapped<_It>(_UNth, _STD move(_Last));
+            _Seek_wrapped(_Nth, _UFinal);
+
+            _Nth_element_common(_Get_unwrapped(_STD move(_First)), _STD move(_UNth), _STD move(_UFinal),
+                _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _Nth;
+        }
+
+        template <random_access_range _Rng, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<iterator_t<_Rng>, _Pr, _Pj>
+        constexpr borrowed_iterator_t<_Rng> operator()(
+            _Rng&& _Range, iterator_t<_Rng> _Nth, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            _Adl_verify_range(_RANGES begin(_Range), _Nth);
+            _Adl_verify_range(_Nth, _RANGES end(_Range));
+            auto _UNth   = _Get_unwrapped(_Nth);
+            auto _UFinal = [&] {
+                if constexpr (common_range<_Rng>) {
+                    return _Uend(_Range);
+                } else if constexpr (sized_range<_Rng>) {
+                    return _RANGES next(_Ubegin(_Range), _RANGES distance(_Range));
+                } else {
+                    return _RANGES next(_UNth, _Uend(_Range));
+                }
+            }();
+            _Seek_wrapped(_Nth, _UFinal);
+
+            _Nth_element_common(
+                _Ubegin(_Range), _STD move(_UNth), _STD move(_UFinal), _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _Nth;
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Pr, class _Pj>
+        static constexpr void _Nth_element_common(_It _First, _It _Nth, _It _Last, _Pr _Pred, _Pj _Proj) {
+            _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+
+            if (_Nth == _Last) {
+                return; // nothing to do
+            }
+
+            while (_ISORT_MAX < _Last - _First) { // divide and conquer, ordering partition containing Nth
+                subrange<_It> _Mid = _RANGES _Partition_by_median_guess_unchecked(_First, _Last, _Pred, _Proj);
+
+                if (_Mid.end() <= _Nth) {
+                    _First = _Mid.end();
+                } else if (_Mid.begin() <= _Nth) {
+                    return; // Nth inside fat pivot, done
+                } else {
+                    _Last = _Mid.begin();
+                }
+            }
+
+            // sort any remainder
+            _RANGES _Insertion_sort_common(_STD move(_First), _STD move(_Last), _STD move(_Pred), _STD move(_Proj));

```

Comment:
I thought after passing something via `_Pass_fn` it would not be necessary to move it again or was it to call `_Pass_fn` again?

Reply:
The latter: it's only necessary to `_Pass_fn` something once. The result of `_Pass_fn` can then be moved and copied freely.